File Formats (Input/Output)
===========================

.. automodule:: formats 

Fasta
-----

Reading fasta
~~~~~~~~~~~~~
is as easy as 

::

  r = FastaReader(file)
  for header, seq in r.get_entries():
      print(header)
      print(seq)

file can either be a path or a file object. 

.. autoclass:: formats.fasta.FastaReader
    :members:

Writing fasta
~~~~~~~~~~~~~    

::

    w = FastaWriter(file, split=60)
    header = "some_random_nucleotides"
    seq = "ACTGACATT"
    w.write_entry(header, seq)
    w.close()

again, file can be a path of a file object. 
``split`` speficies, after how many characters a sequence will be wrapped in multiple lines. Default is 80.


.. autoclass:: formats.fasta.FastaWriter
    :members:

Fast5
-----
Fast5 is a sequence format generated by [Oxford Nanopore](http://nanoporetech.com) devices. This class is designed to read formats up to 
the sequencing kit SQK-MAP006). Compatibility with more recent formats is not guaranteed. 

As it serves for rather exotic use-cases, this file will most likely drop out of future versions of this framework. 

Example
~~~~~~~
::

    f5 = Fast5File('lambda_burnin_ch101_file2_strand.fast5')
    seq = f5.get_seq('template')

.. autoclass:: formats.fast5.Fast5File
    :members: 


